### ISSUE TITLE
dedup -o clobbers existing file too early

### ISSUE THREAD

**Issue by tmtmtmtm** [jqnatividad/qsv#777]
With `data.csv` from https://gist.github.com/tmtmtmtm/2f872e9253057f8b78d21dcfaf35a735, then running

```
qsv fmt --quote-always data.csv |  qsv dedup -o data.csv

``` 

now gives the error:

```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', src/cmd/dedup.rs:175:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

whereas with earlier versions this works fine.

**Comment by tmtmtmtm** [comment]
I'm also seeing the same error with `qsv 0.88.2-mimalloc-apply;fetch;foreach;generate;Luau 0.561;to;self_update-72-72 (x86_64-unknown-linux-gnu compiled with Rust 1.67.1) prebuilt`

**Comment by jqnatividad** [comment]
Hi @tmtmtmtm , I can reproduce this now, but can you tell me why you need to target the same file as output?

If you set `--output` to a new file, or a different  existing file, it works as expected.

Also, I downloaded qsv 0.22.0, and I get the same "index out of bounds" error.

```bash
./qsv22 fmt --quote-always data.csv | ./qsv22 dedup -o data.csv
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', src/cmd/dedup.rs:95:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by tmtmtmtm** [comment]
If it helps you figure out what's going on, it looks like this started working in 0.39, and then stopped working again with 0.88. Everything I've tried between those worked (bisecting; so I haven't tried all versions.)

Maybe I have a different usage pattern to normal, but it doesn't seem strange to me to want to take an input file, remove duplicates from it, and then save that back as the same file.

But if this approach just happened to accidentally work on previous versions, and isn't supported behaviour I've no problem with that: it's easy to pipe through `sponge` instead. Mainly I've raised it in case it's an unnoticed regression or a sign of something else unexpectedly changing, like a filehandle being left open too long or something `</handwave>`.
